parsers/CA_NS.py

```diff
@@ -1,213 +1,198 @@
 #!/usr/bin/env python3
-
-# The datetime library is used to handle datetimes
 from datetime import datetime, timezone
 from logging import Logger, getLogger
+from typing import Any
 
 from requests import Session
 
+from electricitymap.contrib.config import ZoneKey
+from electricitymap.contrib.lib.models.event_lists import (
+    ExchangeList,
+    ProductionBreakdownList,
+)
+from electricitymap.contrib.lib.models.events import ProductionMix
+from parsers.lib.exceptions import ParserException
+
+LOAD_URL = "https://www.nspower.ca/library/CurrentLoad/CurrentLoad.json"
+MIX_URL = "https://www.nspower.ca/library/CurrentLoad/CurrentMix.json"
+PARSER = "CA_NS.py"
+SOURCE = "nspower.ca"
+# Sanity checks: verify that reported production doesn't exceed listed capacity
+# by a lot. In particular, we've seen error cases where hydro production ends
+# up calculated as 900 MW which greatly exceeds known capacity of 418 MW.
+MEGAWATT_LIMITS = {
+    "coal": 1300,
+    "gas": 700,
+    "biomass": 100,
+    "hydro": 500,
+    "wind": 700,
+}
+# This is based on validation logic in
+# https://www.nspower.ca/site/renewables/assets/js/site.js. In practical terms,
+# I've seen hydro production go way too high (>70%) which is way more than
+# reported capacity.
+FRACTION_LIMITS = {
+    # The validation JS reports an error when Solid Fuel (coal) is over 85%,
+    # but as far as I can tell, that can actually be a valid result, I've seen
+    # it a few times. Use 98% instead.
+    "coal": (0, 0.98),
+    "gas": (0, 0.5),
+    "biomass": (0, 0.15),
+    "hydro": (0, 0.60),
+    "wind": (0, 0.55),
+    "imports": (0, 0.50),
+}
+ZONE_KEY = ZoneKey("CA-NS")
+
+
+def _get_ns_info(
+    session: Session, logger: Logger
+) -> (ExchangeList, ProductionBreakdownList):
+    base_loads = session.get(LOAD_URL).json()  # Base loads in MW
+    mixes_percent = session.get(MIX_URL).json()  # Electricity breakdowns in %
+    if any(
+        base_load["datetime"] != mix_percent["datetime"]
+        for base_load, mix_percent in zip(base_loads, mixes_percent)
+    ):
+        raise ParserException(PARSER, "source data is out of sync", ZONE_KEY)
```

Comment:
Did you see [this comment](https://github.com/electricitymaps/electricitymaps-contrib/pull/6050#discussion_r1371097592)? As far as I can tell, the two arrays should always be synchronised; if they're not, it's probably because of a race condition, and I wouldn't trust the data in them at all. I think it would be better to retry the requests rather than falling back to a search.

Reply:
I did but assuming the date times that are returned are always accurate that would not be a problem and it would be better for our internal system if it only raised errors on actual failures.

Reply:
I've never seen the two arrays disagree in practice, so I'm not sure the added complexity of a fallback search would be worthwhile. I think using retries could achieve the same benefit in a simpler way. I'll add a commit to show what I mean; maybe you'll like it!